### PR TITLE
[SPARK-18256] Improve the performance of event log replay in HistoryServer

### PR DESCRIPTION
**Pull request by JoshRosen**
## What changes were proposed in this pull request?

This patch significantly improves the performance of event log replay in the HistoryServer via two simple changes:

- **Don't use `extractOpt`**: it turns out that `json4s`'s `extractOpt` method uses exceptions for control flow, causing huge performance bottlenecks due to the overhead of initializing exceptions. To avoid this overhead, we can simply use our own` Utils.jsonOption` method. This patch replaces all uses of `extractOpt` with `Utils.jsonOption` and adds a style checker rule to ban the use of the slow `extractOpt` method.
- **Don't call `Utils.getFormattedClassName` for every event**: the old code called` Utils.getFormattedClassName` dozens of times per replayed event in order to match up class names in events with SparkListener event names. By simply storing the results of these calls in constants rather than recomputing them, we're able to eliminate a huge performance hotspot by removing thousands of expensive `Class.getSimpleName` calls.

## How was this patch tested?

Tested by profiling the replay of a long event log using YourKit. For an event log containing 1000+ jobs, each of which had thousands of tasks, the changes in this patch cut the replay time in half:

image

Prior to this patch's changes, the two slowest methods in log replay were internal exceptions thrown by `Json4S` and calls to `Class.getSimpleName()`:

image

After this patch, these hotspots are completely eliminated.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 1]
```diff
@@ -540,7 +544,8 @@ private[spark] object JsonProtocol {
 
   def taskStartFromJson(json: JValue): SparkListenerTaskStart = {
     val stageId = (json \ "Stage ID").extract[Int]
-    val stageAttemptId = (json \ "Stage Attempt ID").extractOpt[Int].getOrElse(0)
+    val stageAttemptId =
+      Utils.jsonOption(json \ "Stage Attempt ID").map(_.extract[Int]).getOrElse(0)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: what's the difference? Is `extractOpt` really slow?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: if so maybe we should add a scalastyle to ban it

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: oh wait you already did 😄

**Review by andrewor14 — APPROVED** [review]

**Comment by andrewor14** [comment]
LGTM. That's a massive amount of time spent in `Class.getSimpleName`!

**Comment by andrewor14** [comment]
retest this please

**Comment by yhuai** [comment]
LGTM

**Review by vanzin — APPROVED** [review]

**Comment by yhuai** [comment]
Cool. Merging to master!

**Pull request closed without merging**